### PR TITLE
feat: adds exceptions to discriminate client, server and network errors

### PR DESCRIPTION
**Pull request by antonioturdo**
Throw different types of exceptions to allow to handle them in the most appropriate way

### CODE REVIEW

File: src/ClientAdapter/Psr18Adapter.php [hunk 1]
```diff
@@ -76,9 +78,10 @@ public function makeRequest(string $method, string $path, ?string $body = null):
 
         try {
             return $this->client->sendRequest($request);
-        } catch (ClientExceptionInterface $e) {
-            // is this correct?
-            throw new TransportException($e->getMessage(), 0, $e);
+        } catch (NetworkExceptionInterface $networkException) {
+            throw new TransportException($networkException->getMessage(), 0, $networkException);
```
Comment (Trusted97) [inline src/ClientAdapter/Psr18Adapter.php:82]: Do you think that is possible add a test for cover this new exception?

File: src/ClientAdapter/SymfonyAdapter.php [hunk 2]
```diff
@@ -54,15 +53,17 @@ public function makeRequest(string $method, string $path, ?string $body = null):
...
-            return new ClientResolvedResponse($httpResponse->getStatusCode(), $httpResponse->getContent(false));
-        } catch (TransportExceptionInterface|ClientExceptionInterface|ServerExceptionInterface|RedirectionExceptionInterface $e) {
-            // ClientExceptionInterface, ServerExceptionInterface and RedirectionExceptionInterface never occurs because throw parameter is set to false
+            return new ClientResolvedResponse($httpResponse->getStatusCode(), $httpResponse->getContent());
+        } catch (TransportExceptionInterface $e) {
             throw new TransportException($e->getMessage(), 0, $e);
+        } catch (ServerExceptionInterface $serverException) {
+            throw new ServerException($serverException->getMessage(), 0, $serverException);
+        } catch (ClientExceptionInterface|RedirectionExceptionInterface $clientException) {
+            throw new ClientException($clientException->getMessage(), 0, $clientException);
```
Comment (Trusted97) [inline src/ClientAdapter/SymfonyAdapter.php:66]: Do you think that is possible add a test for cover this new exception?

**Review by JellyBellyDev — COMMENTED** [review]
I believe this is a design theme. DX side, do you think it's better to catch specific `exceptions` or `if` on any status code returned by the response?

**Comment by Trusted97** [comment]
My two cents i think it's better catch specific `exceptions` on any status code returned

**Comment by antonioturdo** [comment]
I agree with @Trusted97, I'd prefer to handle exceptions. I have only one doubt regarding the `BaseResponse` class, at this point the `isSuccess` and `fatalError` properties should be removed, considering that in theory in case of errors an exception should always be raised? Or is this something that we can defer to the future?

**Review by JellyBellyDev — APPROVED** [review]
LGTM 👍

**Pull request merged**